GetIntoTeachingApi/Controllers/TeachingEventsController.cs

```diff
@@ -185,7 +186,22 @@ public async Task<IActionResult> Upsert([FromBody] TeachingEvent teachingEvent)
                 _crm.Save(teachingEvent.Building);
             }
 
-            _crm.Save(teachingEvent);
+            try

```

Comment:
Using exceptions to control flow like this is a smell... I would prefer the logic to live inside the validator, you're shoe-horning the `ModelState` manually here which is a good indicator. I would suggest adding a rule against the `ReadableId` that goes off to the CRM with a method like:

```
public TeachingEvent GetTeachingEvent(string readableId) {}
```

If the CRM returns a matching event then fail the validation.